mod.go

```diff
@@ -0,0 +1,669 @@
+	z  = bits.LeadingZeros64(x[3]); if z <  64 { return }
+	z += bits.LeadingZeros64(x[2]); if z < 128 { return }
+	z += bits.LeadingZeros64(x[1]); if z < 192 { return }
+	z += bits.LeadingZeros64(x[0]); return
+}
+
+func onesCount(x Int) (z int) {
+	z =	bits.OnesCount64(x[0]) +
+		bits.OnesCount64(x[1]) +
+		bits.OnesCount64(x[2]) +
+		bits.OnesCount64(x[3])
+	return
+}
+
+func shiftLeft(x Int, s int) (z Int) {
+	if s < 0 {
+		z = shiftright(x, uint(-s))
+	} else {
+		z = shiftleft(x, uint(s))
+	}
+	return
+}
+
+//func shiftRight(x Int, s int) (z Int) {
+//	if s < 0 {
+//		z = shiftleft(x, uint(-s))
+//	} else {
+//		z = shiftright(x, uint(s))
+//	}
+//	return
+//}
+
+func shiftleft(x Int, s uint) (z Int) {
+	w := s/64	// word shift
+	l := s%64	// left shift
+	r := 64-l	// right shift
+
+	var t Int
+
+	// Bit shift
+
+	t[0] = (x[0] << l)
+	t[1] = (x[1] << l) | (x[0] >> r)
+	t[2] = (x[2] << l) | (x[1] >> r)
+	t[3] = (x[3] << l) | (x[2] >> r)
+
+	// Word shift
+
+	i := 0
+	j := w
+
+	for ; j<4; {
+		z[j] = t[i]
+		i++
+		j++
+	}
+
+	return
+}
+
+func shiftright(x Int, s uint) (z Int) {
+	w := s/64	// word shift
+	r := s%64	// right shift
+	l := 64-r	// left shift
+
+	var t Int
+
+	// Bit shift
+
+	t[0] = (x[0] >> r) | (x[1] << l)
+	t[1] = (x[1] >> r) | (x[2] << l)
+	t[2] = (x[2] >> r) | (x[3] << l)
+	t[3] = (x[3] >> r)
+
+	// Word shift
+
+	i := 0
+	j := w
+
+	for ; j<4; {
+		z[i] = t[j]
+		i++
+		j++
+	}
+
+	return
+}
+
+// shiftright320 shifts a 320-bit value 0-63 bits right
+// z = x >> (s % 64)
+
+func shiftright320(x [5]uint64, s uint) (z [5]uint64) {
+	r := s%64	// right shift
+	l := 64-r	// left shift
+
+	z[0] = (x[0] >> r) | (x[1] << l)
+	z[1] = (x[1] >> r) | (x[2] << l)
+	z[2] = (x[2] >> r) | (x[3] << l)
+	z[3] = (x[3] >> r) | (x[4] << l)
+	z[4] = (x[4] >> r)
+
+	return
+}
+
+// reciprocal computes a 320-bit value representing 1/m
+//
+// Notes:
+// - reciprocal(0) = 0
+// - reciprocal(1) = 2^320-1
+// - otherwise, the result is normalised to have non-zero most significant word
+// - starts with a 32-bit division, refines with newton-raphson iterations
+
+func reciprocal(m Int) (mu [5]uint64) {
+
+	// 0 or a power of 2?
+
+	s := leadingZeros(m)
+
+	if onesCount(m) <= 1 {
+		if s >= 255 {	// m <= 1
+			mu[4] = -m[0]
+			mu[3] = -m[0]
+			mu[2] = -m[0]
+			mu[1] = -m[0]
+			mu[0] = -m[0]
+		} else {
+			mu[4] = 0x8000000000000000 >> ((254-s) & 63)
+		}
+		return
+	}
+
+	// Check for reciprocal in the cache
+
+	cacheIndex := (m[3]^m[2]^m[1]^m[0]) & cacheMask
+
+	cache[cacheIndex].rw.RLock()
+
+	for w:=0; w<cacheWays; w++ {
+		if cache[cacheIndex].mod[w][3] != m[3] { continue }
+		if cache[cacheIndex].mod[w][2] != m[2] { continue }
+		if cache[cacheIndex].mod[w][1] != m[1] { continue }
+		if cache[cacheIndex].mod[w][0] != m[0] { continue }
+
+		for i:=0; i<5; i++ {
+			mu[i] = cache[cacheIndex].inv[w][i]
+		}
+
+		cache[cacheIndex].hit++
+		cache[cacheIndex].rw.RUnlock()
+
+		return
+	}
+
+	cache[cacheIndex].miss++
```

Comment:
Isn't this an unsafe concurrent modification, since you're only holding the readlock ? `x++` is not atomic


(and if so, same a few lines above with `hit++`)

Reply:
Yes, it is unsafe, except it is not very important to have an exact result for these cache statistics.